torch_geometric/nn/models/revgnn.py

```diff
@@ -0,0 +1,287 @@
+        num_groups (Optional[int], optional): The number of groups for
+            group additive reverse.
+        disable (bool, optional): This disables using the InvertibleFunction.
+            Therefore, it executes functions without memory savings.
+        num_bwd_passes (int, optional):
+            Number of backward passes to retain a link with the output.
+            After the last backward pass the output is discarded
+            and memory is freed.
+    """
+    def __init__(
+        self,
+        gnn: Union[nn.Module, nn.ModuleList],
+        split_dim: int = -1,
+        num_groups: Optional[int] = None,
+        disable: bool = False,
+        num_bwd_passes: int = 1,
+    ):
+        super().__init__(disable, num_bwd_passes)
+
+        if isinstance(gnn, nn.ModuleList):
+            n_groups = len(gnn)
+            if num_groups is not None:
+                assert (n_groups == num_groups
+                        ), "Number of GNN groups mismatch with num_group"
+            num_groups = n_groups
+            gnn_groups = gnn
+        else:
+            assert num_groups is not None, "Please specific number of groups"
+            gnn_groups = self._create_groups(gnn, num_groups)
+
+        assert num_groups >= 2, "Number of groups should not be smaller than 2"
+
+        self.gnn_groups = gnn_groups
+        self.split_dim = split_dim
+        self.num_groups = num_groups
+
+    def _forward(self, x: Tensor, edge_index: Adj, *args):
+        xs = torch.chunk(x, self.num_groups, dim=self.split_dim)
+        args_chunks = self._get_arg_chucks(args)
+        y_in = sum(xs[1:])
+        ys = []
+        for i in range(self.num_groups):
+            y_in = xs[i] + self.gnn_groups[i](y_in, edge_index, *
+                                              args_chunks[i])
+            ys.append(y_in)
+
+        y = torch.cat(ys, dim=self.split_dim)
+
+        return y
+
+    def _inverse(self, y: Tensor, edge_index: Adj, *args):
+        ys = torch.chunk(y, self.num_groups, dim=self.split_dim)
+        args_chunks = self._get_arg_chucks(args)
+        xs = []
+        for i in range(self.num_groups - 1, -1, -1):
+            if i != 0:
+                y_in = ys[i - 1]
+            else:
+                y_in = sum(xs)
+            x = ys[i] - self.gnn_groups[i](y_in, edge_index, *args_chunks[i])
+            xs.append(x)
+
+        x = torch.cat(xs[::-1], dim=self.split_dim)
+
+        return x
+
+    def _get_arg_chucks(self, args):
+        if len(args) == 0:
+            args_chunks = [()] * self.num_groups
+        else:
+            chunked_args = list(
+                map(
+                    lambda arg: torch.chunk(arg, self.num_groups, dim=self.
+                                            split_dim)
+                    if arg is not None else [None] * self.num_groups,
+                    args,
+                ))
+            args_chunks = list(zip(*chunked_args))
```

Comment:
The `map()` function could be directly put inside `zip()`? This way we can save an intermediate list instantiation.

Reply:
Good idea.